File: khalti-android/src/main/java/khalti/checkOut/EBanking/EBanking.java
package khalti.checkOut.EBanking;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.AppBarLayout;
import android.support.design.widget.TextInputLayout;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.support.v7.widget.AppCompatTextView;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;
import android.widget.LinearLayout;

import com.jakewharton.rxbinding.view.RxView;
import com.jakewharton.rxbinding.widget.RxTextView;

import java.util.HashMap;
import java.util.List;

import khalti.R;
import khalti.carbonX.widget.Button;
import khalti.carbonX.widget.FrameLayout;
import khalti.carbonX.widget.ProgressBar;
import khalti.checkOut.EBanking.contactForm.ContactFormFragment;
import khalti.checkOut.EBanking.helper.BankAdapter;
import khalti.checkOut.EBanking.helper.BankPojo;
import khalti.checkOut.EBanking.helper.BankingData;
import khalti.utils.EmptyUtil;
import khalti.utils.NetworkUtil;
import khalti.utils.ResourceUtil;
import rx.Observable;

public class EBanking extends Fragment implements EBankingContract.View {

    private RecyclerView rvList;
    private LinearLayout llIndented;
    private ProgressBar pdLoad;
    private AppCompatTextView tvMessage, tvHeader;
    private FrameLayout flTryAgain, flCloseSearch, flSearch;
    private Button btnTryAgain;
    private AppBarLayout appBarLayout;
    private TextInputLayout tilSearch;
    private EditText etSearch;

    private FragmentActivity fragmentActivity;
    private EBankingContract.Presenter presenter;
    private BankAdapter bankAdapter;

    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View mainView = inflater.inflate(R.layout.ebanking, container, false);
        fragmentActivity = getActivity();
        presenter = new EBankingPresenter(this);

        rvList = mainView.findViewById(R.id.rvList);
        llIndented = mainView.findViewById(R.id.llIndented);
        pdLoad = mainView.findViewById(R.id.pdLoad);
        tvMessage = mainView.findViewById(R.id.tvMessage);
        flTryAgain = mainView.findViewById(R.id.flTryAgain);
        flCloseSearch = mainView.findViewById(R.id.flCloseSearch);
        flSearch = mainView.findViewById(R.id.flSearch);
        btnTryAgain = mainView.findViewById(R.id.btnTryAgain);
        appBarLayout = mainView.findViewById(R.id.appBar);
        tilSearch = mainView.findViewById(R.id.tilSearch);
        etSearch = mainView.findViewById(R.id.etSearch);
        tvHeader = mainView.findViewById(R.id.tvHeader);

        presenter.onCreate(NetworkUtil.isNetworkAvailable(fragmentActivity));

        return mainView;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        presenter.onDestroy();
    }

    @Override
    public void toggleIndented(boolean show) {
        llIndented.setVisibility(show ? View.VISIBLE : View.GONE);
    }

    @Override
    public void setUpList(List<BankPojo> bankList) {
        appBarLayout.setVisibility(View.VISIBLE);
        rvList.setVisibility(View.VISIBLE);
        bankAdapter = new BankAdapter(fragmentActivity, bankList);
        rvList.setAdapter(bankAdapter);
        rvList.setHasFixedSize(false);
        RecyclerView.LayoutManager layoutManager = new GridLayoutManager(fragmentActivity, 3);
        rvList.setLayoutManager(layoutManager);
    }

    @Override
    public void showIndentedNetworkError() {
        pdLoad.setVisibility(View.INVISIBLE);
        flTryAgain.setVisibility(View.INVISIBLE);
        tvMessage.setVisibility(View.VISIBLE);
        tvMessage.setText(ResourceUtil.getString(fragmentActivity, R.string.network_error_body));
    }

    @Override
    public void showIndentedError(String error) {
        pdLoad.setVisibility(View.INVISIBLE);
        flTryAgain.setVisibility(View.VISIBLE);
        tvMessage.setVisibility(View.VISIBLE);
        tvMessage.setText(error);
    }

    @Override
    public void openMobileForm(BankingData bankingData) {
        ContactFormFragment contactFormFragment = new ContactFormFragment();
        Bundle bundle = new Bundle();
        bundle.putSerializable("data", bankingData);
        contactFormFragment.setArguments(bundle);
        contactFormFragment.show(getFragmentManager(), contactFormFragment.getTag());
    }

    @Override
    public Observable<HashMap<String, String>> getItemClickObservable() {
        return bankAdapter.getItemClickObservable();
    }

    @Override
    public HashMap<String, Observable<Void>> setOnClickListener() {
        return new HashMap<String, Observable<Void>>() {{
            put("try_again", RxView.clicks(btnTryAgain));
            put("open_search", RxView.clicks(flSearch));
            put("close_search", RxView.clicks(flCloseSearch));
        }};
    }

    @Override
    public Observable<CharSequence> setEditTextListener() {
        return RxTextView.textChanges(etSearch);
    }

    @Override
    public void filterList(String text) {
        fragmentActivity.runOnUiThread(() -> bankAdapter.setFilter(text));
    }

    @Override
    public void flushList() {
        etSearch.setText("");
        bankAdapter.setFilter("");
    }

    @Override
    public void toggleSearch(boolean show) {
        if (show) {
            etSearch.requestFocus();
        }
        flCloseSearch.setVisibility(show ? View.VISIBLE : View.INVISIBLE);
        tilSearch.setVisibility(show ? View.VISIBLE : View.INVISIBLE);
        tvHeader.setVisibility(!show ? View.VISIBLE : View.INVISIBLE);
        flSearch.setEnabled(!show);

        android.widget.FrameLayout.LayoutParams lp = (android.widget.FrameLayout.LayoutParams) flSearch.getLayoutParams();
        lp.gravity = show ? Gravity.CENTER_VERTICAL | Gravity.START : Gravity.CENTER_VERTICAL | Gravity.END;
        flSearch.setLayoutParams(lp);
    }

    @Override
    public void toggleKeyboard(boolean show) {
        InputMethodManager inputManager = (InputMethodManager) fragmentActivity.getSystemService(Context.INPUT_METHOD_SERVICE);
        if (EmptyUtil.isNotNull(inputManager)) {
            if (show) {
                inputManager.showSoftInput(etSearch, InputMethodManager.SHOW_IMPLICIT);
            } else {
                inputManager.hideSoftInputFromWindow(etSearch.getWindowToken(), 0);
            }
        }
    }

    @Override
    public void setPresenter(EBankingContract.Presenter presenter) {
        this.presenter = presenter;
    }
}

File: khalti-android/src/main/java/com/khalti/checkOut/ebanking/EBanking_.kt
package com.khalti.checkOut.ebanking

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.LinearLayout
import androidx.appcompat.widget.AppCompatTextView
import androidx.appcompat.widget.SearchView
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView

import com.google.android.material.appbar.AppBarLayout
import com.google.android.material.button.MaterialButton
import com.jakewharton.rxbinding.support.v7.widget.RxSearchView
import com.jakewharton.rxbinding.view.RxView
import com.khalti.R
import com.khalti.checkOut.ebanking.contactForm.ContactFormFragment
import com.khalti.checkOut.ebanking.helper.BankAdapter
import com.khalti.checkOut.ebanking.helper.BankPojo
import com.khalti.checkOut.ebanking.helper.BankingData
import com.khalti.checkOut.wallet.WalletContract
import com.khalti.signal.Signal
import com.khalti.utils.EmptyUtil
import com.khalti.utils.NetworkUtil
import com.khalti.utils.ResourceUtil
import com.khalti.utils.ViewUtil
import kotlinx.android.synthetic.main.banking.*
import kotlinx.android.synthetic.main.banking.view.*
import kotlinx.android.synthetic.main.wallet_form.view.*

import java.util.HashMap

import rx.Observable
import rx.subjects.PublishSubject

class EBanking_ : Fragment(), EBankingContract.View {

    private lateinit var mainView: View
    private var fragmentActivity: FragmentActivity? = null
    private lateinit var bankAdapter: BankAdapter

    private lateinit var presenter: EBankingContract.Presenter

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        mainView = inflater.inflate(R.layout.banking, container, false)
        fragmentActivity = activity
        presenter = EBankingPresenter_(this)

        presenter.onCreate()

        return mainView
    }

    override fun onDestroy() {
        super.onDestroy()
        presenter.onDestroy()
    }

    override fun toggleIndented(show: Boolean) {
        ViewUtil.toggleView(mainView.llIndented, show)
        ViewUtil.toggleViewInvisible(mainView.flLoad, show)
    }

    override fun toggleSearch(show: Boolean) {
        ViewUtil.toggleView(mainView.flSearchBank, show)
    }

    override fun toggleSearchError(show: Boolean) {
        ViewUtil.toggleView(mainView.rvList, !show)
        ViewUtil.toggleView(mainView.llIndented, show)
        ViewUtil.toggleView(mainView.tvMessage, true)
        ViewUtil.setText(tvMessage, ResourceUtil.getString(fragmentActivity, R.string.no_banks))
    }

    override fun setupList(bankList: MutableList<BankPojo>): Signal<Map<String, String>> {
        ViewUtil.toggleView(mainView.appBar, true)
        ViewUtil.toggleView(mainView.rvList, true)
        bankAdapter = BankAdapter(bankList)
        mainView.rvList.adapter = bankAdapter
        mainView.rvList.setHasFixedSize(false)
        val layoutManager = GridLayoutManager(fragmentActivity, 3)
        mainView.rvList.layoutManager = layoutManager

        return bankAdapter.getClickedItem()
    }

    override fun showIndentedNetworkError() {
        ViewUtil.toggleViewInvisible(mainView.flLoad, false)
        ViewUtil.toggleViewInvisible(mainView.btnTryAgain, false)
        ViewUtil.toggleView(mainView.tvMessage, true)
        ViewUtil.setText(mainView.tvMessage, ResourceUtil.getString(fragmentActivity, R.string.network_error_body))
    }

    override fun showIndentedError(error: String) {
        ViewUtil.toggleViewInvisible(mainView.flLoad, false)
        ViewUtil.toggleView(mainView.btnTryAgain, true)
        ViewUtil.toggleView(mainView.tvMessage, true)
        ViewUtil.setText(mainView.tvMessage, error)
    }

    override fun openMobileForm(bankingData: BankingData) {
        val contactFormFragment = ContactFormFragment()
        val bundle = Bundle()
        bundle.putSerializable("data", bankingData)
        contactFormFragment.arguments = bundle
        if (EmptyUtil.isNotNull(fragmentManager)) {
            contactFormFragment.show(fragmentManager!!, contactFormFragment.tag)
        }
    }

    override fun setOnClickListener(): Map<String, Signal<Any>> = object : HashMap<String, Signal<Any>>() {
        init {
            put("try_again", ViewUtil.setClickListener(mainView.btnTryAgain))
        }
    }

    override fun setSearchListener(): Signal<String> {
        return ViewUtil.setSearchListener(mainView.svBank)
    }

    override fun filterList(text: String): Signal<Int> {
        val signal = Signal<Int>()

        if (EmptyUtil.isNotNull(bankAdapter)) {
            fragmentActivity!!.runOnUiThread {
                val count = bankAdapter.setFilter(text)
                if (EmptyUtil.isNotNull(count)) {
                    signal.emit(count!!)
                }
            }
        }

        return signal
    }

    override fun hasNetwork(): Boolean {
        return NetworkUtil.isNetworkAvailable(fragmentActivity)
    }

    override fun setPresenter(presenter: EBankingContract.Presenter) {
        this.presenter = presenter
    }
}
